### DISCUSSION TITLE
Different colormaps in subplots

### DISCUSSION THREAD

**Discussion by NikitaSakharov** [mtex-toolbox/mtex#907] [Ask Anything] [answered]
Hello, Ralf, Rudiger.

Here is the next issue I faced with. So I need your help again, sorry.

I'm trying to plot mtexFigure, which is made from 4 subplots. I'd like each of them to have (or not to have) it's own colorBar and I don't have a success in it. So here is what I get

This code

mtexFig = newMtexFigure('layout',[2,2]);

nextAxis(1,1)
plot(ebsd, ebsd.bc)
colormap gray

nextAxis(1,2)
plot(grains, grains.meanOrientation)
hold on
plot(gB, 'linecolor','black','linewidth',1)
plot(twinBoundary, 'linecolor','red','linewidth',1)
plot(grains(nonTwinId).boundary, 'linecolor','black','linewidth',2)
hold off

nextAxis(2,1)
plot(grains(nonTwinId), grains(nonTwinId).meanOrientation)
hold on
plot(grains(nonTwinId).boundary,'linecolor','black','linewidth',1)
hold off

nextAxis(2,2)
plot(mergedGrains, twins_param)
mtexColorbar

leads to the following figure
![untitled](https://user-images.githubusercontent.com/62562246/112106909-4f470180-8bbf-11eb-9304-37fc276a3337.jpg)

And in general something is wrong with plotting multiple plots.
If I use the followig code 

mtexFig = newMtexFigure('layout',[2,2]);

nextAxis(1,1)
plot(grains, grains.meanOrientation)
hold on
plot(gB, 'linecolor','black','linewidth',1)
plot(twinBoundary, 'linecolor','red','linewidth',1)
plot(grains(nonTwinId).boundary, 'linecolor','black','linewidth',2)
hold off

nextAxis(1,2)
plot(grains(twinId), grains(twinId).meanOrientation)
hold on
plot(gB, 'linecolor','black','linewidth',1)
plot(twinBoundary, 'linecolor','red','linewidth',1)
hold off

nextAxis(2,1)
plot(mergedGrains, twins_param)
mtexColorbar

nextAxis(2,2)
plot(grains(nonTwinId), grains(nonTwinId).meanOrientation)
hold on
plot(grains(nonTwinId).boundary,'linecolor','black','linewidth',1)
hold off

I get this
![untitled2](https://user-images.githubusercontent.com/62562246/112107695-56224400-8bc0-11eb-8d6f-ab13398ff65d.jpg)

**Comment by kilir** [comment, marked as answer]
Hi Nikita,
here's an example:
```
mtexdata titanium

plot(ebsd)
nextAxis
plot(ebsd,ebsd.iq)
mtexColorMap(gca,'black2white')
nextAxis
plot(ebsd,ebsd.ci)
mtexColorMap(gca,'parula')
mtexColorbar

% get rid of empty colorbars
colorbar(f.children(1),'off')
```
Turning off the unneeded colorbar is not so elegant, maybe there should be an automatic way to find those directly.

Your second result is usually caused by the figure not updating properly. Can you try:
```
f=gcm;
f.drawNow
```

Cheers,
Rüdiger

**Reply by NikitaSakharov** [reply]
Rudiger, many thanks!

Your suggestions helped a lot.
It works well now.

**Comment by NikitaSakharov** [comment]
Unfortunately, now everythings works in a wrong way again. 

Code 

mtexFig = newMtexFigure('layout',[2,2]);
nextAxis(1,1)
plot(ebsd, ebsd.bc)
mtexColorMap(gca,'black2white')

nextAxis(1,2)
plot(grains, grains.meanOrientation)
hold on
plot(gB, 'linecolor','black','linewidth',1)
plot(twinBoundary, 'linecolor','red','linewidth',1)
plot(grains(nonTwinId).boundary, 'linecolor','black','linewidth',2)
hold off

nextAxis(2,1)
plot(mergedGrains, twins_param)
mtexColorbar('title',title)

nextAxis(2,2)
plot(mergedGrains, twins_dens)
mtexColorbar('title','плотность двойниковых границ')

colorbar(mtexFig.children(1),'off')
colorbar(mtexFig.children(2),'off')

leads to the figure without colorbars

![untitled](https://user-images.githubusercontent.com/62562246/112630145-42880f00-8e46-11eb-9cea-19440f7202c0.jpg)

and adding

 f=gcm;
f.drawNow

does nothing.

So what sould I fix? Or where can I try to find the fix or workaround ?

Thank you in advance!

Sincerely, Nikita.

**Reply by kilir** [reply]
Hi,
I think you should skip one of the `mtexColorbar` since the first one turns them on, the second one turns them off again.
Cheers,
Rüdiger

**Reply by NikitaSakharov** [reply]
Rudiger, thanks again!

It works, partially. I mean now I have two colorbars, but I also need to have two different titles for them. It's funny, but I have two identical titles for two subplots. Matlab takes one which is from the last subplot. 

It seems, I want too much ))

Here is a code 

%new figure
mtexFig = newMtexFigure('layout',[2,2]);

nextAxis(1,1)
plot(ebsd, ebsd.bc)
mtexColorMap(gca,'black2white')

nextAxis(1,2)
plot(grains, grains.meanOrientation)
mtexColorbar('title',title) %switch bar on
hold on
plot(gB, 'linecolor','black','linewidth',1)
plot(twinBoundary, 'linecolor','red','linewidth',1)
plot(grains(nonTwinId).boundary, 'linecolor','black','linewidth',2)
hold off

nextAxis(2,1)
plot(mergedGrains, twins_param)
mtexColorbar('title','title') %switch bar off
nextAxis(2,2)
plot(mergedGrains, twins_dens)
mtexColorbar('title',title) %switch bar on again

%and switch off manually some of them
colorbar(mtexFig.children(1),'off')
colorbar(mtexFig.children(2),'off')

And here is what I get
![untitled](https://user-images.githubusercontent.com/62562246/112655724-be438500-8e61-11eb-8935-3d0fe69c8bb2.jpg)

So the colorbars are fixed, colobars' titles are left to be fixed

Many thanks for assistance, Rudiger!

**Reply by kilir** [reply]
Hi Nikita,

you can specify title with the mtexColorbar:

```
mtexdata titanium
f=newMtexFigure
plot(ebsd,ebsd.iq)
mtexColorMap(gca,'black2white')
nextAxis
plot(ebsd)
nextAxis
plot(ebsd,ebsd.ci)
mtexColorMap(gca,'parula')

mtexColorbar('multiple','title',{'iq', 'doesnmatterwillbegone' 'ci'})

% get rid of empty colorbars
pause(1) % I get a funny timing problem here
colorbar(f.children(2),'off')
```

**Reply by NikitaSakharov** [reply]
Rudiger, many thanks, it works now nearly in the way it should.

Here is a code I use

    
mtexFig = newMtexFigure('layout',[2,2]);

nextAxis(1,1)
plot(ebsd, ebsd.bc)
mtexColorMap(gca,'black2white')
mtexColorbar

nextAxis(1,2)
plot(grains, grains.meanOrientation)
hold on
plot(gB, 'linecolor','black','linewidth',1)
plot(twinBoundary, 'linecolor','red','linewidth',1)
plot(grains(nonTwinId).boundary, 'linecolor','black','linewidth',2)
hold off
mtexColorbar('title','title')

nextAxis(2,1)
plot(mergedGrains, twins_param)
mtexColorbar('title','title')

nextAxis(2,2)
plot(mergedGrains, twins_dens)
mtexColorbar('title','плотность двойниковых границ')
mtexColorbar('multiple','title',{'bc map', 'nonTwins are marked','twin param', 'плотность двойниковых границ, мкм^{-1}'})
colorbar(mtexFig.children(1),'off')
colorbar(mtexFig.children(2),'off')

And here is output it gives.

![untitled](https://user-images.githubusercontent.com/62562246/112833059-d7854500-909e-11eb-87fe-3480dc2d6146.jpg)
